4365: scaled line width

alexkiri opened the pull request:
A quick solution for https://github.com/TrenchBroom/TrenchBroom/issues/4365

These are the things I've tested and seem to be scaled correctly with the current solution:
- selected / deselected brush edges
- selected / deselected entity edges
- selection alignment helpers
- clip tool path
- vertex / edge / face tool hover highlight
- entity links
- xyz axes for map views
- brush tool path
- portal file edges
- point file path
- rotate/scale/shear controls for map views
- face inspector controls

The GridRenderer is intentionally not updated, as it looks fine without line width scaling, in my opinion.

Comparison
3D View https://imgsli.com/MjE5MzI5
face inspector https://imgsli.com/MjIzMjMy

kduske commented:
Why did you create this PR if it's not ready to be reviewed?

alexkiri commented:
I created it in order to have a discussion related to some specific parts of the code. 
So, the first question would be: is it ok to store the pixelRatio in Preferences? Or is there any other singleton functionality?
Ad the second question, is this approach ok? I didn't get too deep into the code yet, but for the part that's not working (face inspector controls) I'll have to understand the logic better.

kduske commented:
But I thought the idea was to set the line width automatically without a preference. And there is still the problems that this solution using glLineWidth doesn't work on some OpenGL implementations.

alexkiri commented:
Yes, it _is_ set automatically. I only used the preferences to store the pixelRatio value retrieved in `TrenchBroomApp.cpp`, thinking that it's going to be needed in different places, and trying to avoid to import `QApplication` and `QScreen` in those places. But currently, there's only one file where the logic is needed. I can cleanup the implementation and remove the Preferences and use the pixelRatio logic directly in the `PrimitiveRenderer.cpp`. But this will not make the face inspector fix easier.

And regarding the troublesome OpenGL implementations, how does the issue manifest? Is the `glLineWidth ` ignored? In this case, this PR will not affect them in any way

kduske commented:
Ok, I see. I'll take a closer look. Thanks for the info!

alexkiri commented:
I have removed the Preferences workaround, and moved the entire logic inside a single file `PrimitiveRenderer.cpp`. This works OK for most cases (enumerated in the PR description). Making it work for the face inspector controls will need more changes, and will porbably go in a different PR

kduske commented on common/src/Renderer/PrimitiveRenderer.cpp:
```diff
@@ -75,7 +78,9 @@ bool PrimitiveRenderer::LineRenderAttributes::operator<(
   IndexRangeRenderer& renderer, ActiveShader& shader) const
 {
-  glAssert(glLineWidth(m_lineWidth));
+  QScreen* screen = QApplication::screens().at(0);
```
It would be better to use `QWindow::devicePixelRatio` because you are just making a guess as to which monitor TB's window will be on. The renderer should also not depend on Qt.

A better solution would be to pass `devicePixelRatio` into `RenderContext`. `RenderContext` is constructed in `MapViewBase::doRender`, and there, you can access `QWindow::devicePixelRatio` to use the correct value for the view being rendered.

alexkiri commented on common/src/View/MapViewBase.cpp:
```diff
@@ -1031,6 +1032,8 @@ void MapViewBase::doRender()
   renderContext.setShowFog(pref(Preferences::ShowFog));
   renderContext.setShowGrid(grid.visible());
   renderContext.setGridSize(grid.actualSize());
+  const auto ratio = QApplication::primaryScreen()->devicePixelRatio(); // to be changed
```
Still using `primaryScreen()` for now, as I have no idea how to get the main QWindow. @kduske ideas?
I have tried:
```
const auto window1 = QApplication::focusWindow(); // NULL when using file picker
const auto window2 = QApplication::activeWindow(); // NULL when using file picker
const auto window3 = QWindow::fromWinId(0); // works OK, but am I doing the right thing?
```

kduske replied:
Don't use `QScreen` at all. instead call `window()->devicePixelRatio();`

kduske replied:
Oh, and I think you can just inline the `ratio` variable.

kduske commented on common/src/Renderer/PrimitiveRenderer.cpp:
```diff
@@ -367,11 +367,12 @@ void PrimitiveRenderer::renderLines(RenderContext& renderContext)
 {
   ActiveShader shader(renderContext.shaderManager(), Shaders::VaryingPUniformCShader);
 
+  const auto scaledLineWidth = static_cast<float>(renderContext.dpiScale());
```
This variable name is wrong. I would also suggest not to cast here and to store `dpiScale` as `float` in `RenderContext`.

kduske commented on common/src/Renderer/PrimitiveRenderer.cpp:
```diff
@@ -73,9 +73,9 @@ bool PrimitiveRenderer::LineRenderAttributes::operator<(
 
 void PrimitiveRenderer::LineRenderAttributes::render(
-  IndexRangeRenderer& renderer, ActiveShader& shader) const
+  IndexRangeRenderer& renderer, ActiveShader& shader, float dpiScale) const
```
`dpiScale` can be made `const` here.

kduske commented on common/src/Renderer/RenderContext.h:
```diff
@@ -74,6 +74,7 @@ class RenderContext
 
   bool m_showGrid;
   FloatType m_gridSize;
+  FloatType m_dpiScale;
```
This should really be `float` because that's what `glLineWidth` takes.

alexkiri commented:
I have added dpi scaling for face inspector controls. But I have found some little things
The `UVOriginTool` and `UVScaleTool` tools use a vey thin line width, that seems too small, even when applying the dpi scaling logic. More specifically, these
```
UVOriginTool.cpp
edgeRenderer.renderOnTop(renderBatch, 0.25f);

UVScaleTool.cpp
handleRenderer.render(renderBatch, color, 0.5f);
```
![image](https://github.com/TrenchBroom/TrenchBroom/assets/6061770/deeb6cff-9ed5-4960-93d8-0ad04e110f41)
The red vertical and horizontal controls are barely visible. Should these be adjusted?

kduske commented:
> I have added dpi scaling for face inspector controls. But I have found some little things The `UVOriginTool` and `UVScaleTool` tools use a vey thin line width, that seems too small, even when applying the dpi scaling logic. More specifically, these
> 
> ```
> UVOriginTool.cpp
> edgeRenderer.renderOnTop(renderBatch, 0.25f);
> 
> UVScaleTool.cpp
> handleRenderer.render(renderBatch, color, 0.5f);
> ```
> 
> ![image](https://private-user-images.githubusercontent.com/6061770/286863018-deeb6cff-9ed5-4960-93d8-0ad04e110f41.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDEzMzczMjEsIm5iZiI6MTcwMTMzNzAyMSwicGF0aCI6Ii82MDYxNzcwLzI4Njg2MzAxOC1kZWViNmNmZi05ZWQ1LTQ5NjAtOTNkOC0wYWQwNGUxMTBmNDEucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQUlXTkpZQVg0Q1NWRUg1M0ElMkYyMDIzMTEzMCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyMzExMzBUMDkzNzAxWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NzNkNjViZmNiZjRhYmM1NWY4NzBmNDdiMGNkNTQ3MmMxMTBjYWZkYjJlMWE0MjJmNzg3MWY3ZjQ0YjI5OTQzNSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ._dSMjkNu1VupajNjhTsPYhcPTj3JQvKb-bUDHjs1j3Y) Edit: before / after comparison https://imgsli.com/MjIzMjMy The red vertical and horizontal controls are barely visible. Should these be adjusted?
> 
> Also, for the `UVView` I added the dpi scaling logic into the fragment shader. I could do that for the GridRenderer as well, but I intentionally didn't. The reason is that it actually looks fine without scaling (thinner on denser displays) and more importantly, it's not interactable / selectable. The `UVView` lines _need_ to be scaled because they are interactable / used as controls. Let me know what you think about that @kduske

I agree with most of your decisions, except that now the grid line handles in the UV view are too thick, esp. when compared to the red highlights. They should have the same width if possible - maybe that can't be done though because they are drawn by different means? Maybe it would be better to move the code to do the red highlights into the shader then. Otherwise, feel free to adjust the line widths to something sensible.

I have just tested this on my mac laptop, and it looks good! Maybe the brush edges are even a bit too prominent now, but we can adjust this.

kduske commented:
It looks better, but the grid lines are still too thick on mac. But I can fix that in a followup. Thanks for addressing my feedback!

alexkiri commented:
> It looks better, but the grid lines are still too thick on mac. But I can fix that in a followup. Thanks for addressing my feedback!

You're welcome. I don't have a mac, so I can't easily test without any virtual machine shenanigans.
If you get to it, be make sure to check the portal file rendering, it uses ticker line width.
